src/graphics/vulkan.zig

```diff
@@ -0,0 +1,150 @@
+	const T = @typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child;
+	var count: u32 = 0;
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, null}));
+	const list = allocator.alloc(T, count);
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, list.ptr}));
+	return list;
+}
+
+pub fn enumerateInstanceLayerProperties(allocator: NeverFailingAllocator) []c.VkLayerProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceLayerProperties, allocator, .{});
+}
+
+pub fn enumerateInstanceExtensionProperties(allocator: NeverFailingAllocator, layerName: ?[*:0]u8) []c.VkExtensionProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceExtensionProperties, allocator, .{layerName});
+}
+
+pub const Instance = struct { // MARK: Instance
+	var instance: c.VkInstance = undefined;
+
+	const validationLayers: []const [*:0]const u8 = &.{
+		"VK_LAYER_KHRONOS_validation",
+	};
+
+	fn checkValidationLayerSupport() bool {
+		const availableLayers = enumerateInstanceLayerProperties(main.stackAllocator);
+		defer main.stackAllocator.free(availableLayers);
+		for(validationLayers) |layerName| continueOuter: {
+			for(availableLayers) |layerProperties| {
+				if(std.mem.eql(u8, std.mem.span(layerName), std.mem.span(@as([*:0]const u8, @ptrCast(&layerProperties.layerName))))) {
+					break :continueOuter;
+				}
+			}
+			std.log.warn("Couldn't find validation layer {s}", .{layerName});
+			return false;
+		}
+		return true;
+	}
+
+	pub fn init() void {
+		if(c.gladLoaderLoadVulkan(null, null, null) == 0) {
+			@panic("GLAD failed to load Vulkan functions");
+		}
+		const appInfo = c.VkApplicationInfo{
+			.sType = c.VK_STRUCTURE_TYPE_APPLICATION_INFO,
+			.pApplicationName = "Cubyz",
+			.applicationVersion = c.VK_MAKE_VERSION(0, 0, 0),
+			.pEngineName = "custom",
+			.engineVersion = c.VK_MAKE_VERSION(0, 0, 0),
+			.apiVersion = c.VK_API_VERSION_1_0,
+		};
+		var glfwExtensionCount: u32 = 0;
+		const glfwExtensions: [*c][*c]const u8 = c.glfwGetRequiredInstanceExtensions(&glfwExtensionCount);
+
+		const availableExtensions = enumerateInstanceExtensionProperties(main.stackAllocator, null);
+		defer main.stackAllocator.free(availableExtensions);
+		std.log.debug("Availabe vulkan instance extensions:", .{});
+		for(availableExtensions) |ext| {
+			std.log.debug("\t{s}", .{@as([*:0]const u8, @ptrCast(&ext.extensionName))});
+		}
+
+		const createInfo = c.VkInstanceCreateInfo{
+			.sType = c.VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
+			.pApplicationInfo = &appInfo,
+			.enabledExtensionCount = glfwExtensionCount,
+			.ppEnabledExtensionNames = glfwExtensions,
+			.ppEnabledLayerNames = validationLayers.ptr,
+			.enabledLayerCount = if(checkValidationLayerSupport()) validationLayers.len else 0,
```

Comment:
If all you do is toggle, you can just try creating the instance w/ layers and if it fails with `VK_ERROR_LAYER_NOT_PRESENT` do it again w/o layers. What you are doing would make sense if you were to dynamically drop layers which are not available.

Reply:
But we request one layer.
Anyway, since we already have this code in place, I suppose we can keep it.